### ISSUE TITLE
Add save and load of patterns

### ISSUE THREAD

**Issue by jbphet** [phetsims/quake#18]
The 0.0.1 release didn't have the ability to save and load patterns.  This functionality is important for sharing patterns and putting them into specs, so it should be added in the next release.

**Comment by jbphet** [comment]
This is turning out to be way harder than I'd anticipated.  Documentation is spotty and inconsistent for file manipulation, and the behavior doesn't seem to be consistent across platforms.  Through a lot of searching and experimentation, I was able to get something working on the emulator, but it doesn't work on my Samsung S7 tablet.  I decided to check the code in anyway because I need to work on other things.

I also implemented support on the browser, but I haven't tested it much because I don't think people are planning to use it much.  So, as of this writing, it exists, but is only very superficially tested.

**Comment by jbphet** [comment]
As of now, file saving works fine on the emulator and the S7 tablet, and the file open dialog works on both as well.  The problem occurs when trying to read the file on the S7.  A very non-helpful error is thrown that says:

```
"Uncaught TypeError: Cannot convert undefined or null to object", source: https://localhost/cordova.js (314)
```

This doesn't occur on the emulator, so I'm not at all sure what the problem is or how to correct it.  I have to leave it for now and continue it later.

**Comment by jbphet** [comment]
I spent a fair amount of time trying to get file loading to work on the S7 tablet and so far I haven't gotten it to work.  Through exploring the cordova code, popping up alert messages, and setting breakpoints in the Java code in Android Studio, I was able to determine that the root cause is a file permissions error when trying to read the file.  I tried adding some new permissions requests to the `config.xml` file, but that didn't fix the problem.

I've decided to move forward with a release anyway.  My hope is that this permissions thing is unique to the situation where I'm debugging an app via Android Studio, and it won't be a problem in the version of the app distributed through the Play Store.  This is admittedly optimistic, but there's another reason to do this: Users don't actually need to be able to load files for this feature to be somewhat useful.  They can still _save_ patterns and send them to developers, they just can't re-load them once they do.  So, onward with the release.

**Comment by jbphet** [comment]
Over Slack, @jessegreenberg reported:

> "Save" and "Load" are working well for me on my Galaxy s8! I saved two different patterns and switched back and forth between them by loading each. Very cool!

Good news.  Hopefully this will be the case for the other users as well.

**Comment by jbphet** [comment]
@BLFiedler has reported that loading files does not work on his Pixel 4a running Android 12.  Bummer.

**Comment by jbphet** [comment]
Another user has reported that file load for patterns isn't working, so I should do some investigation.

**Comment by jbphet** [comment]
I tried adding a line to the `config.xml` file to request internal read permissions, since I thought that might be the issue.  The line looked like this:

```xml
        <config-file target="AndroidManifest.xml" parent="/*">
            <uses-permission android:name="android.permission.READ_INTERNAL_STORAGE" />
        </config-file>
```

...and I had to add the following to the root element:

```
  xmlns:android="http://schemas.android.com/apk/res/android
```

This did indeed update the `AndroidManifest.xml` file, but it didn't solve the problem.  I still can't load files on the Samsung S7.

**Comment by jbphet** [comment]
I just tried using a plugin that is supposed to allow code-based querying and modification of the app's permissions.  The plugin is called `cordova-plugin-android-permissions`, and I followed the information on [its npm page](https://www.npmjs.com/package/cordova-plugin-android-permissions) to set it up and try it out.  I started out by querying whether the app had `READ_EXTERNAL_STORAGE` and `WRITE_EXTERNAL_STORAGE` permissions and logging the result to the console, which I can see when running on the S7 via Android Studio.  According to my experiments with this plugin, the app has read permissions but not write permissions, which is exactly the opposite of what I expected based on the behavior that I'm seeing.  Then, since it was saying that it didn't have write permissions, I added code to request it, which appeared to succeed based on the return code.  However, it didn't change the behavior - I still can't load pattern files from the app when running on the Tab S7.

This is all quite confusing, since the permissions that are read through the plugin don't seem to match up with the problems that I'm seeing.  Also, I tried changing the requested permissions on the `config.xml` file, and that didn't seem to affect the permissions read by the plugin.  So, right now, this is not adding up to a clear picture that can help me resolve the issue.

I didn't commit the code for reading the permissions, and I don't think I will, since it turned out to be a dead end and it increases complexity.  I have some other ideas of where to go from here, and I can always rewrite this if I really need to.

**Comment by jbphet** [comment]
I discussed the issue that I'm having with loading files on the Samsung S7 and other Android platforms with @mattpen today, and got some ideas of things to pursue.  He asked to see exactly the error that I'm running into when I step through the Java code in case he could lend some insight based on that.  So, below is the text of that error.  This was obtained by running the app on the Samsung S7 using Android Studio and setting breakpoints in the Java code.

```
e = {FileNotFoundException@22386} "java.io.FileNotFoundException: /storage/emulated/0/Download/file____data_user_0_edu.colorado.phet.hapticsPlayground_files_pattern.json: open failed: EACCES (Permission denied)"
 backtrace = {Object[13]@22391} 
 cause = {ErrnoException@22392} "android.system.ErrnoException: open failed: EACCES (Permission denied)"
  errno = 13
  functionName = "open"
  backtrace = {Object[18]@22401} 
  cause = {ErrnoException@22392} "android.system.ErrnoException: open failed: EACCES (Permission denied)"
   errno = 13
   functionName = "open"
   backtrace = null
   cause = {ErrnoException@22392} "android.system.ErrnoException: open failed: EACCES (Permission denied)"
   detailMessage = null
   stackTrace = {StackTraceElement[17]@22403} 
   suppressedExceptions = {Collections$EmptyList@22395}  size = 0
   shadow$_klass_ = {Class@15815} "class android.system.ErrnoException"
   shadow$_monitor_ = 0
  detailMessage = null
  stackTrace = {StackTraceElement[17]@22403} 
  suppressedExceptions = {Collections$EmptyList@22395}  size = 0
  shadow$_klass_ = {Class@15815} "class android.system.ErrnoException"
  shadow$_monitor_ = 0
 detailMessage = "/storage/emulated/0/Download/file____data_user_0_edu.colorado.phet.hapticsPlayground_files_pattern.json: open failed: EACCES (Permission denied)"
 stackTrace = {StackTraceElement[12]@22397} 
 suppressedExceptions = {Collections$EmptyList@22395}  size = 0
 shadow$_klass_ = {Class@16347} "class java.io.FileNotFoundException"
```

Here is the stack trace for this error.  This is more for my reference than for @mattpen's.

```
readFileAs:1141, FileUtils (org.apache.cordova.file)
run:331, FileUtils$5 (org.apache.cordova.file)
run:690, FileUtils$25 (org.apache.cordova.file)
runWorker:1167, ThreadPoolExecutor (java.util.concurrent)
run:641, ThreadPoolExecutor$Worker (java.util.concurrent)
run:920, Thread (java.lang)
```

Also for my reference, to get to this point I set breakpoints in `CordovaBridge.java`, then `PluginManager.exec`, then `FileUtils.execute`, then `FileUtils.readFileAs`, then `FileSystem.readFileAtURL`.

Assigning to @mattpen to have a look.

**Comment by jbphet** [comment]
FWIW, I just tried adding the following line to the `config.xml` file based on [something I read in the file plugin documentation](https://cordova.apache.org/docs/en/11.x/reference/cordova-plugin-file/index.html#android-quirks).  It didn't seem to make any difference - file loading still errors out on the Samsung S7.

```
<preference name="AndroidPersistentFileLocation" value="Internal" />
```

**Comment by mattpen** [comment]
@jbphet - have you reviewed this guide: https://developer.android.com/training/data-storage/shared/documents-files?

I think you're already requesting READ_EXTERNAL_STORAGE, but without it you can't access files in shared storage.

In particular that guide has a section on how to "persist permissions": https://developer.android.com/training/data-storage/shared/documents-files#persist-permissions

**Comment by mattpen** [comment]
https://developer.android.com/training/data-storage/manage-all-files

Maybe you also need to request MANAGE_EXTERNAL_STORAGE permission???

**Comment by jbphet** [comment]
I did some experiments and found that I could save and subsequently load files out of the app's local storage area, as opposed to the shared "Downloads" file that it was using before.  That's great, but I couldn't get the save dialog and the "chooser" (for opening previously created files) to point to this portion of the file system, so I basically redesigned to UI to not have any dialogs and to support an explicit "Export" capability where the user can copy and paste a pattern.  I think this will serve the needed purpose.  I need to make a few more refinements, and then I'll review it with the team at the next meeting.

**Comment by jbphet** [comment]
The save/load functionality is implemented and was published in the v1.0.0 release.  It no longer uses dialogs for saving and loading files (other than a quick "Save succeeded" alert dialog).  This seems to work on the Samsung S7 and I suspect it will work on other devices as well.

I'll have the team check it over once we resume meetings again.  It's not perfect, but it gets the job done, and I suspect it is sufficient for our needs at this point.

Here is what the current interface looks like in the browser version:

![image](https://user-images.githubusercontent.com/3082831/176288168-c2595e97-83ae-4aaf-850a-8ffadc95af2c.png)

I'm going to defer this until we review the changes and will either close the issue or follow up on further changes, depending on the result.

**Comment by jbphet** [comment]
Changes reviewed and approved.  Closing.
